File: templates/problems/07_trees/06_subtree/typescript/solution.ts
import { TreeNode } from "./helpers";

function sameTree(p: TreeNode | null, q: TreeNode | null): boolean {
    if (p === null && q === null) {
        return true;
    }
    if (p === null || q === null || p.val !== q.val) {
        return false;
    }
    return sameTree(p.left, q.left) && sameTree(p.right, q.right);
}

export function solveIsSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
    if (root === null) {
        return subRoot === null;
    }
    if (sameTree(root, subRoot)) {
        return true;
    }
    return solveIsSubtree(root.left, subRoot) || solveIsSubtree(root.right, subRoot);
}

File: templates/problems/07_trees/06_subtree/python/solution.py
from helpers import TreeNode, Optional


def solve_is_subtree(root: Optional[TreeNode], sub_root: Optional[TreeNode]) -> bool:
    """Return true if sub_root is a subtree of root.

    Time: O(m*n), Space: O(h).
    """
    if root is None:
        return sub_root is None
    if _same_tree(root, sub_root):
        return True
    return solve_is_subtree(root.left, sub_root) or solve_is_subtree(root.right, sub_root)


def _same_tree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    if p is None and q is None:
        return True
    if p is None or q is None or p.val != q.val:
        return False
    return _same_tree(p.left, q.left) and _same_tree(p.right, q.right)
